### ISSUE TITLE
Question regarding setting a code value

### ISSUE THREAD

**Issue by kdschlosser** [newAM/monitorcontrol#3]
I am running on Windows and wanting to get and set the backlight of my monitor. It is using the legacy VCP code. How does one go about doing this?

the VCP code is 0x13

get
MH = color choice
ML = returned value

set
SH = color choice
SL = new value

color choices are
white = 0x00
red = 0x01
green = 0x02
blue = 0x03

The set should be pretty easy to do with some bit shifting.
The get on the other hand I am not sure how I would go about passing the color choice

any help would be appreciated.

**Comment by newAM** [comment]
I am thinking of maybe `dataclass` for the WRGB for the return value & argument.

```python
import dataclasses


@dataclasses.dataclass
class WRGB:
    white: Optional[int] = None
    red: Optional[int] = None
    green: Optional[int] = None
    blue: Optional[int] = None
```

Where a `None` value would imply a zero color-choice.

I do not have time right now, but I can implement this later if you want.

**Comment by kdschlosser** [comment]
OK let me give you some background of what I am attempting to do. There is no single do it all module for python to interface with all things video related. meaning. I use Windows so for my example I am only going to be referring to Windows .

Windows API controls for the Monitor
Windows API calls for all others, setting the resolution and arranging the monitors are examples.
DDC/CI control of a monitor
parsing EDID data
Video card driver control (direct)

I have ported the entire ATI Display SDK to Python. This allows for dealing with the video card drivers directly.

I am about 50-60%  finished with the port for for the Nvidia SDK
The EDID parser is complete
I have all of the DDC/CI specification completed for the commands not the tables.

I am sure that there are API functions in the Nvidia SDK as well as the ATI SDK that would probably do what I need I wanted this project to be complete in terms of abilities that are not vendor specific. so someone with an Intel video card would still be able to use all of the "standard" things.

It is a pretty chunky project at some 54000 lines of code so far. It is pure python up to the API functions for Windows, the ATI ASK and the Nvidia SDK. ctypes is what I am using to interface with the shared libraries.

The legacy back light command is a single command to query all colors. the color to be queried needs to be sent along with the vcp code to the monitor. the color choice is sent as the MH byte (byte 3) in the get command. The Windows API function GetVCPFeatureAndVCPFeatureReply only has 2 input parameters the handle to the monitor and the vcp code. There is no place from what I can see to pass any additional data to the monitor inside that get command. I am able to use the set command no problem, just need to use bit wise operations to set the color choice and the new value into a single python integer and pass that to the SetVCPFeature function.

My issues lies in getting the current value. The non legacy backlight command(s) are not a single vcp code. there is a code for each color. So there is no issue with obtaining the individual color values because all that needs to be passed is the vcp code.

I ask the question for information. you may have come across a way to do this. You do not need to add anything to your library unless you want to. I would have used your library as a start off point if it had be written to work with python 2.7,

**Comment by newAM** [comment]
Ah, I missed that part of the problem entirely.

I took another look at the spec and I see the problem; I have no idea what the outcome of a read would be from the description.

> The SH byte defines whether operation should be performed as a white adjustment or as a red / green or blue backlight adjustment when these are separate light sources (e.g. LEDs).

One guess, this might be a W vs RGB option, and that the value is either W or RGB.

Other than that I would probably go about empirically testing it, set all to zero, increase W, read data, reset, repeat for another value.  Either that or email VESA for clarification, but I am not sure how they would handle a request from a hobbyist.

**Comment by kdschlosser** [comment]
OK so I wasn't going coo coo and you see the issue as well. I do not have a monitor available to use that has this type of control using the pre 2.1 specification. So I personally have no way of testing to see if it works or not. I guess what I am going to do is have the set portion available and maybe not the get for the time being. when someone that possibly has a monitor that uses the legacy style commands comes along I can mess about with it to see if I can get it working. 

I am sure there is a way to also collect the information from the either the Nvidia or ATI drivers if all else fails.

I would bet that VESA changed the standard in 2.1 to the separate colors because of this exact problem and that there is no mechanism to query the display for the information using the Windows API.

I guess I really wanted to make sure I was not going looney tunes.

I manage a project called EventGhost it is used by a couple hundred thousand people so I do not think me asking them would necessarily fall into the hobbyist category. So they may answer the question. The 2.0 standard is pretty old and there may never be a need to ask them.

I do appreciate your time in looking into it and confirming my thoughts on it. I didn't know i you may have stumbled across something that could be used to get the information or not.

**Comment by newAM** [comment]
Oh, EventGhost!  I used that with Windows to connect my monitors to my home automation system, very handy.

The deprecation was in MCSS 2.2, January 26, 2009.  I have an old (2004) NEC LCD1560V+, but on interrogation it doesn't support 0x13.
```
$ sudo ddcutil detect
Display 1
   I2C bus:             /dev/i2c-0
   Supports DDC:        true
   EDID synopsis:
      Mfg id:           NEC
      Model:            NEC LCD1560V+
      Serial number:    43305611YA
      Manufacture year: 2004
      EDID version:     1.3
   VCP version:         Detection failed

$ sudo ddcutil capabilities --display 1
MCCS version: 1.1
VCP Features:
   Feature: 04 (Restore factory defaults)
   Feature: 06 (Restore factory geometry defaults)
   Feature: 08 (Restore color defaults)
   Feature: 0E (Clock)
   Feature: 10 (Brightness)
   Feature: 12 (Contrast)
   Feature: 14 (Select color preset)
      Values:
         02: Display Native
         04: 5000 K
         05: 6500 K
         06: 7500 K
         07: 8200 K
         08: 9300 K
   Feature: 16 (Video gain: Red)
   Feature: 18 (Video gain: Green)
   Feature: 1A (Video gain: Blue)
   Feature: 1E (Auto setup)
   Feature: 20 (Horizontal Position)
   Feature: 30 (Vertical Position)
   Feature: 3E (Clock phase)
   Feature: 68 (unrecognized feature)
      sValues: 01 02 03 04 05 06 07 (interpretation unavailable)
   Feature: A8 (unrecognized feature)
   Feature: B4 (Source Timing Mode)
   Feature: B6 (Display technology type)
   Feature: D6 (Power mode)
   Feature: E3 (manufacturer specific feature)
   Feature: FA (manufacturer specific feature)
   Feature: FF (manufacturer specific feature)
```

I think this feature was used mainly by CRTs, and I got rid of all mine long ago.  Hopefully VESA will be able to give an answer!

**Comment by kdschlosser** [comment]
I have 3 displays from 07 and 09 that are 2.1 but they do not support 0x13 either.. I am going to skip over it for now as it is pretty old anyhow.

The Nvidia SDK port is a bear. I wrote a program for converting Windows SDK header and idl files to python. It works fairly well. I used it on the Nvidia SDK. it got most of the structures and enums correct and also the preprocessor macros but it didn't do so hot with the functions and there are a ton of them. there are also a bunch of com interfaces in the SDK that i have to clean up because of how they were declared in the SDK.

So you are familiar with EventGhost. I ended up being the administrator of it about 3 years ago, I really don't know how I am the one that ended up in the role, but I did. I have been doing my best to extend it's functionality. there are almost 400 plugins available for it now. I use a single copy of EventGhost to run my house. 4 TV's, 1 projector, 6 tuner mcard cable tuner, 5 HTPCs, central PVR, central video game server with over 5000 games, movie library with close to 3000 movies, 5 zone HVAC (3 furnaces, and 2 radiant heating systems), 5 IR remotes, geofencing, 50 zwave dimmers and switches, 30 security sensors, 10 security cameras, zwave door locks, 3 furnace filter monitors, 2 sump pump controllers, flood sensors, 2 fireplaces, landscape lighting. OSD menu systems for the TV's, 5 touch pads, and a bunch more...
